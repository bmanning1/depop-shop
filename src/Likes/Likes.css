.likes-list {
    border: black solid 1px;
    padding: 0;
    width: max-content;
    max-width: 15rem;
    position: absolute;
    top: 1.6rem;
    right: -1rem;
    background-color: white;
    z-index: 100;
    font-size: 1rem;
}

.likes-list li {
    display: flex;
    justify-content: space-between;
    margin: 1rem 1.5rem;
}

.likes-list li button {
    background: none;
    border: none;
    margin: 0 0 0 1rem;
    padding: 0;
    font-size: 0.8rem;
}

.likes-list li button:hover {
    text-decoration: underline;
    cursor: pointer;
}

.likes-count {
    display: block;
    border: none;
    font-weight: 900;
    font-size: 1rem;
    cursor: pointer;
    background-color: red;
}

.likes-count:hover {
    transition: ease-in;
    text-decoration: underline;
}

.likes-count:disabled {
    color: black;
}

.likes-count:focus, .likes-list li button:focus {
    outline: none;
}

.likes {
    position: relative;
}
