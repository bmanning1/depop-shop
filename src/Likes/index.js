import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import './Likes.css';

export const LikesList = ({ likedItems, unlikeItem }) => (
    <ul className="likes-list">
        {likedItems.map(({ id, title }) => (
            <li key={id}>
                <b>{title}</b> <button onClick={() => unlikeItem(id)}>X</button>
            </li>
        ))}
    </ul>
);

export const LikeCount = ({ likes, onClick }) => (
    <button className="likes-count" disabled={likes === 0} onClick={onClick}>
        {`${likes} Like${likes === 1 ? '' : 's'}`}
    </button>
);

const Likes = ({ likedItems, unlikeItem }) => {
    const [showList, setShowList] = useState(false);

    useEffect(() => {
        likedItems.length === 0 && setShowList(false)
    }, [likedItems]);

    return (
        <div className="likes">
            <LikeCount likes={likedItems.length} onClick={() => setShowList(!showList)} />
            {showList && likedItems.length !== 0 &&
                <LikesList likedItems={likedItems} unlikeItem={unlikeItem} />}
        </div>
    )
};

Likes.propTypes = {
    likedItems: PropTypes.arrayOf(PropTypes.exact({
        id: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired
    })).isRequired,
    unlikeItem: PropTypes.func.isRequired
};

export default Likes;