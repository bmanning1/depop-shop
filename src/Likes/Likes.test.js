import React from 'react';
import { shallow } from 'enzyme';
import Likes, { LikeCount, LikesList } from '.';

describe('Likes:', () => {
    let baseComponent;
    const likedItems = [
        {
            id: 1,
            title: 'jumper'
        },
        {
            id: 2,
            title: 'trainers'
        },
    ];

    beforeEach(() => {
        baseComponent = shallow(
            <Likes likedItems={likedItems} unlikeItem={jest.fn()} />
        );
    });

    test('should match snapshot', () => {
        expect(baseComponent).toMatchSnapshot();
    });

    describe('when the like count is clicked', () => {
        test('should show LikesList if there are liked items', () => {
            const component = shallow(
                <Likes likedItems={likedItems} unlikeItem={jest.fn()} />
            );
            component.find('LikeCount').simulate('click');
            expect(baseComponent).toMatchDiffSnapshot(component);
        });

        test('should not show LikesList if there are no liked items', () => {
            const component = shallow(
                <Likes likedItems={[]} unlikeItem={jest.fn()} />
            );
            component.find('LikeCount').simulate('click');
            expect(baseComponent).toMatchDiffSnapshot(component);
        });

        test('should hide LikesList if it is already shown', () => {
            baseComponent.find('LikeCount').simulate('click');
            expect(baseComponent.find('LikesList')).toHaveLength(1);
            baseComponent.find('LikeCount').simulate('click');
            expect(baseComponent.find('LikesList')).toHaveLength(0);
        });
    })
});

describe('LikeCount:', () => {
    let baseComponent, onClick;
    const likes = 2;

    beforeEach(() => {
        onClick = jest.fn();
        baseComponent = shallow(
            <LikeCount likes={likes} onClick={onClick} />
        );
    });

    test('should match snapshot', () => {
        expect(baseComponent).toMatchSnapshot();
    });

    test('should disable the button if there are no likes', () => {
        const component = shallow(
            <LikeCount likes={0} onClick={onClick} />
        );
        expect(baseComponent).toMatchDiffSnapshot(component);
    });

    test('should call the onClick function when the button is clicked', () => {
        baseComponent.find('.likes-count').simulate('click');
        expect(onClick).toHaveBeenCalled();
    });

    test('should display the correct grammer if there is 1 like', () => {
        const component = shallow(
            <LikeCount likes={1} onClick={onClick} />
        );
        expect(baseComponent).toMatchDiffSnapshot(component);
    });
});

describe('LikesList:', () => {
    let baseComponent, unlikeItem;
    const id1 = 1;
    const likedItems = [
        {
            id: id1,
            title: 'jumper'
        },
        {
            id: 2,
            title: 'trainers'
        },
    ];

    beforeEach(() => {
        unlikeItem = jest.fn();
        baseComponent = shallow(
            <LikesList likedItems={likedItems} unlikeItem={unlikeItem} />
        );
    });

    test('should match snapshot', () => {
        expect(baseComponent).toMatchSnapshot();
    });

    test('should unlike the selected item on button click', () => {
        baseComponent.find('li').first().find('button').simulate('click');
        expect(unlikeItem).toHaveBeenCalledWith(id1);
    })
});