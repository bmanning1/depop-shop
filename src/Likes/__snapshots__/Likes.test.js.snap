// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LikeCount: should disable the button if there are no likes 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <button
    className=\\"likes-count\\"
-   disabled={false}
+   disabled={true}
    onClick={[Function mockConstructor]}
  >
-   2 Likes
+   0 Likes
  </button>"
`;

exports[`LikeCount: should display the correct grammer if there is 1 like 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <button
    className=\\"likes-count\\"
    disabled={false}
    onClick={[Function mockConstructor]}
  >
-   2 Likes
+   1 Like
  </button>"
`;

exports[`LikeCount: should match snapshot 1`] = `
<button
  className="likes-count"
  disabled={false}
  onClick={[MockFunction]}
>
  2 Likes
</button>
`;

exports[`Likes: should match snapshot 1`] = `
<div
  className="likes"
>
  <LikeCount
    likes={2}
    onClick={[Function]}
  />
</div>
`;

exports[`Likes: when the like count is clicked should not show LikesList if there are no liked items 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <div
    className=\\"likes\\"
  >
    <LikeCount
-     likes={2}
+     likes={0}
      onClick={[Function onClick]}
    />
  </div>"
`;

exports[`Likes: when the like count is clicked should show LikesList if there are liked items 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -3,6 +3,21 @@
  >
    <LikeCount
      likes={2}
      onClick={[Function onClick]}
    />
+   <LikesList
+     likedItems={
+       Array [
+         Object {
+           \\"id\\": 1,
+           \\"title\\": \\"jumper\\",
+         },
+         Object {
+           \\"id\\": 2,
+           \\"title\\": \\"trainers\\",
+         },
+       ]
+     }
+     unlikeItem={[Function mockConstructor]}
+   />
  </div>"
`;

exports[`LikesList: should match snapshot 1`] = `
<ul
  className="likes-list"
>
  <li
    key="1"
  >
    <b>
      jumper
    </b>
     
    <button
      onClick={[Function]}
    >
      X
    </button>
  </li>
  <li
    key="2"
  >
    <b>
      trainers
    </b>
     
    <button
      onClick={[Function]}
    >
      X
    </button>
  </li>
</ul>
`;
